File: Java/RecipeSharingAPI/src/main/java/microsoft/ghcp/RecipeSharingAPI/Controllers/RecipeController.java
package microsoft.ghcp.RecipeSharingAPI.Controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import microsoft.ghcp.RecipeSharingAPI.Models.Recipe;
import microsoft.ghcp.RecipeSharingAPI.Services.RecipeService;
import java.util.List;

@RestController
@RequestMapping("/api/Recipe")
public class RecipeController {

    @Autowired
    private RecipeService recipeService;

    @GetMapping("/{id}")
    public ResponseEntity<Recipe> getRecipeById(@PathVariable Long id) {
        Recipe recipe = recipeService.getRecipeById(id);
        if (recipe != null) {
            return ResponseEntity.ok(recipe);
        } else {
            return ResponseEntity.noContent().build();
        }
    }

    @GetMapping
    public List<Recipe> getRecipes() {
        return recipeService.getRecipes();
    }

    @GetMapping("/search")
    public List<Recipe> findByTitle(@RequestParam String title) {
        return recipeService.findByTitle(title);
    }

    @GetMapping("/pagination")
    public Page<Recipe> getRecipesWithPagination(@RequestParam int page, @RequestParam int size) {

        Page<Recipe> recipes = recipeService.getRecipesWithPagination(size, page);

        if (recipes.isEmpty()) {
            return Page.empty();
        }

        return recipeService.getRecipesWithPagination(page, size);
    }

    @PostMapping
    public Recipe createRecipe(@RequestBody Recipe recipe) {
        return recipeService.addRecipe(recipe);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Recipe> updateRecipe(@PathVariable Long id, @RequestBody Recipe recipeDetails) {
        Recipe updatedRecipe = recipeService.updateRecipe(id, recipeDetails);
        if (updatedRecipe != null) {
            return ResponseEntity.ok(updatedRecipe);
        } else {
            return ResponseEntity.noContent().build();
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteRecipe(@PathVariable Long id) {
        Recipe recipe = recipeService.getRecipeById(id);

        if (recipe == null) {
            return ResponseEntity.noContent().build();
        }

        recipeService.deleteRecipe(id);

        return ResponseEntity.ok().build();
    }
}

File: C#/RecipeSharingAPI/Controllers/RecipeController.cs
using Microsoft.AspNetCore.Mvc;
using RecipeSharingAPI.Models;
using RecipeSharingAPI.Services;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace RecipeSharingAPI.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class RecipeController : ControllerBase
    {
        private readonly IRecipeService _recipeService;

        public RecipeController(IRecipeService recipeService)
        {
            _recipeService = recipeService;
        }

        [HttpGet]
        public  ActionResult<IEnumerable<Recipe>> GetRecipes()
        {
            var recipes = _recipeService.GetRecipes();
            return Ok(recipes);
        }

        [HttpGet("{id}")]
        public ActionResult<Recipe> GetRecipeById(int id)
        {
            var recipe =  _recipeService.GetRecipe(id);
            if (recipe == null)
            {
                return NotFound();
            }
            return Ok(recipe);
        }

        [HttpGet("search")]
        public ActionResult<IEnumerable<Recipe>> SearchRecipeByTitle(string title)
        {
            var recipes = _recipeService.SearchRecipeByTitle(title);

            if (recipes.Count == 0)
            {
                return NoContent();
            }

            return Ok(recipes);
        }

        [HttpPost]
        public ActionResult<Recipe> CreateRecipe([FromBody] Recipe recipe)
        {
            _recipeService.CreateRecipe(recipe);
            return CreatedAtAction(nameof(GetRecipeById), new { id = recipe.Id }, recipe);
        }

        [HttpPut("{id}")]
        public async Task<IActionResult> UpdateRecipe([FromBody] Recipe recipe)
        {
           
            var existingRecipe = _recipeService.GetRecipe(recipe.Id);
            
            if (existingRecipe == null)
            {
                return NotFound();
            }

            var updatedRecipe = _recipeService.UpdateRecipe(recipe);
            return Ok(updatedRecipe);
        }

        [HttpDelete("{id}")]
        public IActionResult DeleteRecipe(int id)
        {
            var recipe = _recipeService.GetRecipe(id);
            if (recipe == null)
            {
                return NotFound();
            }

            _recipeService.DeleteRecipe(id);
            return NoContent();
        }
    }
}